compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/AggregatedOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,245 @@
+        Value rhs = convertScalarToDtype(b, loc, args[1], args[2].getType(),
+                                         /*isUnsignedCast=*/false);
+        Value mul = b.create<arith::MulFOp>(loc, lhs, rhs);
+        Value add = b.create<arith::AddFOp>(loc, mul, args[2]);
+        b.create<linalg::YieldOp>(loc, add);
+      });
+
+  return genericOp.getResult(0);
+}
+
+// Compute output = exp2(output - input)
+static Value computeSubAndExp2(OpBuilder &builder, Location loc,
+                               AffineMap inputMap, AffineMap outputMap,
+                               Value input, Value output) {
+  SmallVector<AffineMap> compressedMaps =
+      compressUnusedDims(SmallVector<AffineMap>{inputMap, outputMap});
+  inputMap = compressedMaps[0];
+  outputMap = compressedMaps[1];
+
+  SmallVector<utils::IteratorType> iteratorTypes(inputMap.getNumDims(),
+                                                 utils::IteratorType::parallel);
+  auto genericOp = builder.create<linalg::GenericOp>(
+      loc, output.getType(), input, output,
+      SmallVector<AffineMap>{inputMap, outputMap}, iteratorTypes,
+      [&](OpBuilder &b, Location loc, ValueRange args) {
+        // Convert input to the same datatype as output.
+        Value in = convertScalarToDtype(b, loc, args[0], args[1].getType(),
+                                        /*isUnsignedCast=*/false);
+        Value diff = b.create<arith::SubFOp>(loc, args[1], in);
+        Value weight = b.create<math::Exp2Op>(loc, diff);
+        b.create<linalg::YieldOp>(loc, weight);
+      });
+  return genericOp.getResult(0);
+}
+
+AffineMap dropResultsContainingDims(AffineMap map, ArrayRef<int64_t> dims) {
+  SmallVector<int> resultsToDrop;
+  for (auto [idx, result] : llvm::enumerate(map.getResults())) {
+    for (int dim : dims) {
+      if (result.isFunctionOfDim(dim)) {
+        resultsToDrop.push_back(idx);
+        break;
+      }
+    }
+  }
+
+  // Iterate in reverse to preserve indices to drop.
+  for (int result : llvm::reverse(resultsToDrop)) {
+    map = map.dropResult(result);
+  }
+  return map;
+}
+
+FailureOr<SmallVector<Value>>
+OnlineAttentionOp::decomposeOperation(OpBuilder &b) {
+  Location loc = getLoc();
+  Value query = getQuery();
+  Value key = getKey();
+  Value value = getValue();
+  Value oldAcc = getOutput();
+  Value oldMax = getMax();
+  Value oldSum = getSum();
+  Type elementType = getQuery().getType().getElementType();
+
+  FailureOr<AttentionOpDetail> maybeOpInfo =
+      AttentionOpDetail::get(getIndexingMapsArray());
+  assert(succeeded(maybeOpInfo) && "Invalid attention indexing maps");
+  AttentionOpDetail opInfo = maybeOpInfo.value();
+
+  SmallVector<OpFoldResult> sizes = llvm::map_to_vector(
+      getIterationDomain(b), [](Range x) { return x.size; });
+
+  // Since we use exp2 for attention instead of the original exp, we have to
+  // multiply the scale by log2(e). We use exp2 instead of exp as most GPUs
+  // have better support for exp2.
```

Comment:
The comment is interesting. I spent some time on the study and I think it's applied to most of platforms, but not only GPUs. One of my theory is that we are representing values in binary numeral system, and the below stackoverflow link has more explanation. So I think we can replace `GPUs` with `platforms` in the comment.

https://stackoverflow.com/questions/30222836/should-exp2-be-faster-than-exp

Reply:
I'll write platforms and write that we have verified this for some gpus. Thanks for checking!